EpiAware/src/epimodel.jl

```diff
@@ -2,91 +2,117 @@ abstract type AbstractEpiModel end
 
 struct EpiData{T <: Real, F <: Function}
     gen_int::Vector{T}
-    delay_int::Vector{T}
-    delay_kernel::SparseMatrixCSC{T, Integer}
-    cluster_coeff::T
     len_gen_int::Integer
-    len_delay_int::Integer
-    time_horizon::Integer
     transformation::F
 
     #Inner constructors for EpiData object
-    function EpiData(
-            gen_int,
-            delay_int,
-            cluster_coeff,
-            time_horizon::Integer,
-            transformation::Function
-    )
+    function EpiData(gen_int,
+            transformation::Function)
         @assert all(gen_int .>= 0) "Generation interval must be non-negative"
-        @assert all(delay_int .>= 0) "Delay interval must be non-negative"
         @assert sum(gen_int)≈1 "Generation interval must sum to 1"
-        @assert sum(delay_int)≈1 "Delay interval must sum to 1"
 
-        K = generate_observation_kernel(delay_int, time_horizon)
-
-        new{eltype(gen_int), typeof(transformation)}(
-            gen_int,
-            delay_int,
-            K,
-            cluster_coeff,
+        new{eltype(gen_int), typeof(transformation)}(gen_int,
             length(gen_int),
-            length(delay_int),
-            time_horizon,
-            transformation
-        )
+            transformation)
     end
 
-    function EpiData(
-            gen_distribution::ContinuousDistribution,
-            delay_distribution::ContinuousDistribution,
-            cluster_coeff,
-            time_horizon::Integer;
+    function EpiData(gen_distribution::ContinuousDistribution;
             D_gen,
-            D_delay,
             Δd = 1.0,
-            transformation::Function = exp
-    )
+            transformation::Function = exp)
         gen_int = create_discrete_pmf(gen_distribution, Δd = Δd, D = D_gen) |>
                   p -> p[2:end] ./ sum(p[2:end])
-        delay_int = create_discrete_pmf(delay_distribution, Δd = Δd, D = D_delay)
 
-        return EpiData(gen_int, delay_int, cluster_coeff, time_horizon, transformation)
+        return EpiData(gen_int, transformation)
     end
 end
 
-struct DirectInfections <: AbstractEpiModel
+struct DirectInfections{S <: Sampleable} <: AbstractEpiModel
     data::EpiData
+    initialisation_prior::S
 end
 
-function (epimodel::DirectInfections)(_It, init)
-    epimodel.data.transformation.(init .+ _It)
+struct ExpGrowthRate{S <: Sampleable} <: AbstractEpiModel
+    data::EpiData
+    initialisation_prior::S
 end
 
-struct ExpGrowthRate <: AbstractEpiModel
+struct Renewal{S <: Sampleable} <: AbstractEpiModel
     data::EpiData
+    initialisation_prior::S
 end
 
-function (epimodel::ExpGrowthRate)(rt, init)
-    init .+ cumsum(rt) .|> exp
+"""
+    function (epimodel::Renewal)(recent_incidence, Rt)
+
+Compute new incidence based on recent incidence and Rt.
+
+This is a callable function on `Renewal` structs, that encodes new incidence prediction
```

Comment:
Its pretty expressive and gets tested?

Reply:
my point is that you need to read the source to know. For a fairly simple function this might be fine but as things get complex its going to get harder and harder to track things.

Reply:
I think we aren't making progress on this and it seems like a language difference potentially.